pkg/storegateway/storegatewaypb/custom.go

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package storegatewaypb
+
+import (
+	"context"
+
+	"github.com/grafana/dskit/grpcutil"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+)
+
+// customStoreGatewayClient is a custom client which wraps the real gRPC client and re-maps well known
+// gRPC errors into standard golang errors.
+type customStoreGatewayClient struct {
+	wrapped StoreGatewayClient
+}
+
+func NewCustomStoreGatewayClient(cc *grpc.ClientConn) StoreGatewayClient {
+	return &customStoreGatewayClient{
+		wrapped: NewStoreGatewayClient(cc),
+	}
+}
+
+// Series implements StoreGatewayClient.
+func (c *customStoreGatewayClient) Series(ctx context.Context, in *storepb.SeriesRequest, opts ...grpc.CallOption) (StoreGateway_SeriesClient, error) {
+	client, err := c.wrapped.Series(ctx, in, opts...)
+	if err != nil {
+		return client, mapError(err)
+	}
+
+	return &customSeriesClient{client}, nil
+}
+
+// LabelNames implements StoreGatewayClient.
+func (c *customStoreGatewayClient) LabelNames(ctx context.Context, in *storepb.LabelNamesRequest, opts ...grpc.CallOption) (*storepb.LabelNamesResponse, error) {
+	res, err := c.wrapped.LabelNames(ctx, in, opts...)
+	return res, mapError(err)
+}
+
+// LabelValues implements StoreGatewayClient.
+func (c *customStoreGatewayClient) LabelValues(ctx context.Context, in *storepb.LabelValuesRequest, opts ...grpc.CallOption) (*storepb.LabelValuesResponse, error) {
+	res, err := c.wrapped.LabelValues(ctx, in, opts...)
+	return res, mapError(err)
+}
+
+type customSeriesClient struct {
+	StoreGateway_SeriesClient
+}
+
+func (c *customSeriesClient) Recv() (*storepb.SeriesResponse, error) {
+	res, err := c.StoreGateway_SeriesClient.Recv()
+	return res, mapError(err)
+}
+
+func mapError(err error) error {
+	switch {
+	case err == nil:
+		return nil
+	case grpcutil.IsCanceled(err):
+		return context.Canceled
+	case grpcutil.ErrorToStatusCode(err) == codes.DeadlineExceeded:
+		return context.DeadlineExceeded
```

Comment:
If we return `context.Canceled` or `context.DeadlineExceeded` here, then we'll drop any extra information in the gRPC error message, like the [message](https://pkg.go.dev/google.golang.org/grpc@v1.60.0/internal/status#Status.Message). 

What if we instead returned a custom error type here that wraps the gRPC status information, but returns `true` for `errors.Is(err, context.Canceled)` (or `context.DeadlineExceeded`)?

Reply:
This is a very good feedback. I'm proposing a better wrapping here: https://github.com/grafana/mimir/pull/6934/commits/8417cb263e7164991e252398712b18964dcab905

Reply:
Post-merge suggestion: what if we used or extended the type from [dskit's `cancellation` package](https://github.com/grafana/dskit/blob/main/cancellation/error.go), rather than creating another one here?